### PR TITLE
feat!(c++): Refactor code structure to consolidate include and src into a single src directory

### PR DESCRIPTION
**Pull request by acezen**
### Reason for this PR

Our current project structure separates header files and source files into include and src directories, respectively. This approach, while beneficial in some contexts, adds complexity and overhead for smaller projects, **and not fits to protobuf generated code**. To simplify the structure and enhance maintainability, we need to consolidate all code into a single src directory.

### What changes are included in this PR?

Here are some notable changes:
- Consolidate the header and source code into single `src` directory.
- Add subdirectory to classify the `arrow` and `high-level` related code.
- Add `api` headers to let users to easily to include the function of c++ library.

### Are these changes tested?

Yes

### Are there any user-facing changes?

**BREAKING CHANGE:**
- the header path and header name may change, like:
  - `graphar/arrow_chunk_reader.h` to `graphar/arrow/chunk_reader.h`
  - `graphar/arrow_chunk_writer.h` to `graphar/arrow/chunk_writer.h`
- The `copy` way writer function has been moved from `arrow_chunk_writer.h` to `chunk_info_writer.h`

### CODE REVIEW

**Comment by acezen** [comment]
Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
- I use `high-level` for our high level reader and writer directory name
- Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
- And also the `api` 's header names.

**Comment by lixueclaire** [comment]
> Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> 
> * I use `high-level` for our high level reader and writer directory name
> * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> * And also the `api` 's header names.

Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.

**Comment by acezen** [comment]
> > Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> > 
> > * I use `high-level` for our high level reader and writer directory name
> > * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> > * And also the `api` 's header names.
> 
> Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.

You means we should split reader and writer to `info`, `arrow` and `high_level`and also split the same level reader/writer to different directory?

**Comment by lixueclaire** [comment]
> > > Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> > > 
> > > * I use `high-level` for our high level reader and writer directory name
> > > * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> > > * And also the `api` 's header names.
> > 
> > 
> > Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.
> 
> You mean we should split reader and writer to `info`, `arrow` and `high_level`and also split the same level reader/writer to different directory?

I'm talking about the file names under the api directory.

**Comment by acezen** [comment]
> > > > Hi, @lixueclaire, I create this PR to consolidate all code into a single src directory and brings some `BREAKING CHANGE`, may need your opinion about the changes, especially the naming of the directory and file:
> > > > 
> > > > * I use `high-level` for our high level reader and writer directory name
> > > > * Since we use `arrow` folder to separate the arrow reader and writer, so I change the `arrow_chunk_reader` and `arrow_chunk_writer` to `chunk_reader` and `chunk_writer`.
> > > > * And also the `api` 's header names.
> > > 
> > > 
> > > Hi, I think the APIs maybe a little confusing now. Originally, we have `chunk_info_reader` and `arrow_chunk_reader` and high-level graph collections. If we use `basic_reader`, it is not clear what is read. I suggest naming `info_reader/meta_reader`,`arrow_reader/chunk_reader` and `high_level_reader/graph_reader`.
> > 
> > 
> > You mean we should split reader and writer to `info`, `arrow` and `high_level`and also split the same level reader/writer to different directory?
> 
> I'm talking about the file names under the api directory.

I have changed the naming under api, could you help reviewing again?

File: cpp/examples/low_level_reader_example.cc [hunk 1]
```diff
@@ -23,8 +23,7 @@
 #include "arrow/filesystem/api.h"
 
 #include "./config.h"
-#include "graphar/api.h"
-#include "graphar/reader/chunk_info_reader.h"
+#include "graphar/api/basic_reader.h"
```
Comment (lixueclaire) [inline cpp/examples/low_level_reader_example.cc:26]: This need to be updated.

Comment (acezen) [reply cpp/examples/low_level_reader_example.cc]: Thanks for pointing out, fixed.

**Review by lixueclaire — APPROVED** [review]
LGTM now. And it's essential to update our documentation accordingly.

**Pull request merged**